windows10, compile mmcv-full error, focal_loss_cuda.obj : error LNK2001: unresolved external symbol

The error as below:
focal_loss_cuda.obj : error LNK2001: unresolved external symbol "public: long __cdecl at::Tensor::item<long>(void)const " (??$item@J@Tensor@at@@QEBAJXZ)

Then I open the file focal_loss_cuda.cu, find there have four long:
AT_ASSERTM(target.max().item<**long**>() <= (**long**)num_classes
AT_ASSERTM(target.max().item<**long**>() <= (**long**)num_classes

I change all "long" to "int64_t" as below. Now I can install the mmcv-full successfully.

AT_ASSERTM(target.max().item<int64_t>() <= (int64_t)num_classes
AT_ASSERTM(target.max().item<int64_t>() <= (int64_t)num_classes

--- a/README.md
+++ b/README.md
@@ -131,6 +131,19 @@ e.g.,
 CC=clang CXX=clang++ CFLAGS='-stdlib=libc++' MMCV_WITH_OPS=1 pip install -e .
 ```
 
+If you are on Windows10, set the following environment variable before the installing command.
+
+```bash
+set MMCV_WITH_OPS=1
+```
+
+e.g.,
+
+```bash
+set MMCV_WITH_OPS=1
+pip install -e .
+```
+
 Note: If you would like to use `opencv-python-headless` instead of `opencv-python`,
 e.g., in a minimum container environment or servers without GUI,
 you can first install it before installing MMCV to skip the installation of `opencv-python`.

--- a/mmcv/ops/csrc/pytorch/focal_loss_cuda.cu
+++ b/mmcv/ops/csrc/pytorch/focal_loss_cuda.cu
@@ -14,13 +14,11 @@ void SigmoidFocalLossForwardCUDAKernelLauncher(Tensor input, Tensor target,
   cudaStream_t stream = at::cuda::getCurrentCUDAStream();
   AT_DISPATCH_FLOATING_TYPES_AND_HALF(
       input.scalar_type(), "sigmoid_focal_loss_forward_cuda_kernel", [&] {
-        sigmoid_focal_loss_forward_cuda_kernel<scalar_t> << <GET_BLOCKS(
-                                                                 output_size),
-                                                             THREADS_PER_BLOCK,
-                                                             0, stream>>>
-            (output_size, input.data_ptr<scalar_t>(),
-             target.data_ptr<int64_t>(), weight.data_ptr<scalar_t>(),
-             output.data_ptr<scalar_t>(), gamma, alpha, num_classes);
+        sigmoid_focal_loss_forward_cuda_kernel<scalar_t>
+            <<<GET_BLOCKS(output_size), THREADS_PER_BLOCK, 0, stream>>>(
+                output_size, input.data_ptr<scalar_t>(),
+                target.data_ptr<int64_t>(), weight.data_ptr<scalar_t>(),
+                output.data_ptr<scalar_t>(), gamma, alpha, num_classes);
       });
 
   AT_CUDA_CHECK(cudaGetLastError());
@@ -38,13 +36,11 @@ void SigmoidFocalLossBackwardCUDAKernelLauncher(Tensor input, Tensor target,
   cudaStream_t stream = at::cuda::getCurrentCUDAStream();
   AT_DISPATCH_FLOATING_TYPES_AND_HALF(
       input.scalar_type(), "sigmoid_focal_loss_backward_cuda_kernel", [&] {
-        sigmoid_focal_loss_backward_cuda_kernel<scalar_t> << <GET_BLOCKS(
-                                                                  output_size),
-                                                              THREADS_PER_BLOCK,
-                                                              0, stream>>>
-            (output_size, input.data_ptr<scalar_t>(),
-             target.data_ptr<int64_t>(), weight.data_ptr<scalar_t>(),
-             grad_input.data_ptr<scalar_t>(), gamma, alpha, num_classes);
+        sigmoid_focal_loss_backward_cuda_kernel<scalar_t>
+            <<<GET_BLOCKS(output_size), THREADS_PER_BLOCK, 0, stream>>>(
+                output_size, input.data_ptr<scalar_t>(),
+                target.data_ptr<int64_t>(), weight.data_ptr<scalar_t>(),
+                grad_input.data_ptr<scalar_t>(), gamma, alpha, num_classes);
       });
 
   AT_CUDA_CHECK(cudaGetLastError());
@@ -63,13 +59,11 @@ void SoftmaxFocalLossForwardCUDAKernelLauncher(Tensor softmax, Tensor target,
   cudaStream_t stream = at::cuda::getCurrentCUDAStream();
   AT_DISPATCH_FLOATING_TYPES_AND_HALF(
       softmax.scalar_type(), "softmax_focal_loss_forward_cuda_kernel", [&] {
-        softmax_focal_loss_forward_cuda_kernel<scalar_t> << <GET_BLOCKS(
-                                                                 output_size),
-                                                             THREADS_PER_BLOCK,
-                                                             0, stream>>>
-            (output_size, softmax.data_ptr<scalar_t>(),
-             target.data_ptr<int64_t>(), weight.data_ptr<scalar_t>(),
-             output.data_ptr<scalar_t>(), gamma, alpha, num_classes);
+        softmax_focal_loss_forward_cuda_kernel<scalar_t>
+            <<<GET_BLOCKS(output_size), THREADS_PER_BLOCK, 0, stream>>>(
+                output_size, softmax.data_ptr<scalar_t>(),
+                target.data_ptr<int64_t>(), weight.data_ptr<scalar_t>(),
+                output.data_ptr<scalar_t>(), gamma, alpha, num_classes);
       });
 
   AT_CUDA_CHECK(cudaGetLastError());
@@ -85,34 +79,32 @@ void SoftmaxFocalLossBackwardCUDAKernelLauncher(Tensor softmax, Tensor target,
   int output_size = buff.numel();
   at::cuda::CUDAGuard device_guard(grad_input.device());
   cudaStream_t stream = at::cuda::getCurrentCUDAStream();
-  AT_DISPATCH_FLOATING_TYPES_AND_HALF(grad_input.scalar_type(),
-                                      "softmax_focal_loss_backward_cuda1_"
-                                      "kernel",
-                                      [&] {
-    softmax_focal_loss_backward_cuda1_kernel<scalar_t> << <GET_BLOCKS(
-                                                               output_size),
-                                                           THREADS_PER_BLOCK, 0,
-                                                           stream>>>
-        (output_size, softmax.data_ptr<scalar_t>(), target.data_ptr<int64_t>(),
-         weight.data_ptr<scalar_t>(), buff.data_ptr<scalar_t>(), gamma, alpha,
-         num_classes);
-  });
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+      grad_input.scalar_type(),
+      "softmax_focal_loss_backward_cuda1_"
+      "kernel",
+      [&] {
+        softmax_focal_loss_backward_cuda1_kernel<scalar_t>
+            <<<GET_BLOCKS(output_size), THREADS_PER_BLOCK, 0, stream>>>(
+                output_size, softmax.data_ptr<scalar_t>(),
+                target.data_ptr<int64_t>(), weight.data_ptr<scalar_t>(),
+                buff.data_ptr<scalar_t>(), gamma, alpha, num_classes);
+      });
 
   AT_CUDA_CHECK(cudaGetLastError());
 
   output_size = grad_input.numel();
-  AT_DISPATCH_FLOATING_TYPES_AND_HALF(grad_input.scalar_type(),
-                                      "softmax_focal_loss_backward_cuda2_"
-                                      "kernel",
-                                      [&] {
-    softmax_focal_loss_backward_cuda2_kernel<scalar_t> << <GET_BLOCKS(
-                                                               output_size),
-                                                           THREADS_PER_BLOCK, 0,
-                                                           stream>>>
-        (output_size, softmax.data_ptr<scalar_t>(), target.data_ptr<int64_t>(),
-         buff.data_ptr<scalar_t>(), grad_input.data_ptr<scalar_t>(),
-         num_classes);
-  });
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+      grad_input.scalar_type(),
+      "softmax_focal_loss_backward_cuda2_"
+      "kernel",
+      [&] {
+        softmax_focal_loss_backward_cuda2_kernel<scalar_t>
+            <<<GET_BLOCKS(output_size), THREADS_PER_BLOCK, 0, stream>>>(
+                output_size, softmax.data_ptr<scalar_t>(),
+                target.data_ptr<int64_t>(), buff.data_ptr<scalar_t>(),
+                grad_input.data_ptr<scalar_t>(), num_classes);
+      });
 
   AT_CUDA_CHECK(cudaGetLastError());
 }

--- a/mmcv/ops/csrc/pytorch/focal_loss_cuda.cu
+++ b/mmcv/ops/csrc/pytorch/focal_loss_cuda.cu
@@ -8,17 +8,19 @@ void SigmoidFocalLossForwardCUDAKernelLauncher(Tensor input, Tensor target,
                                                const float alpha) {
   int output_size = output.numel();
   int num_classes = input.size(1);
-  AT_ASSERTM(target.max().item<long>() <= (long)num_classes,
+  AT_ASSERTM(target.max().item<int64_t>() <= (int64_t)num_classes,
              "target label should smaller or equal than num classes");
   at::cuda::CUDAGuard device_guard(input.device());
   cudaStream_t stream = at::cuda::getCurrentCUDAStream();
   AT_DISPATCH_FLOATING_TYPES_AND_HALF(
       input.scalar_type(), "sigmoid_focal_loss_forward_cuda_kernel", [&] {
-        sigmoid_focal_loss_forward_cuda_kernel<scalar_t>
-            <<<GET_BLOCKS(output_size), THREADS_PER_BLOCK, 0, stream>>>(
-                output_size, input.data_ptr<scalar_t>(),
-                target.data_ptr<int64_t>(), weight.data_ptr<scalar_t>(),
-                output.data_ptr<scalar_t>(), gamma, alpha, num_classes);
+        sigmoid_focal_loss_forward_cuda_kernel<scalar_t> << <GET_BLOCKS(
+                                                                 output_size),
+                                                             THREADS_PER_BLOCK,
+                                                             0, stream>>>
+            (output_size, input.data_ptr<scalar_t>(),
+             target.data_ptr<int64_t>(), weight.data_ptr<scalar_t>(),
+             output.data_ptr<scalar_t>(), gamma, alpha, num_classes);
       });
 
   AT_CUDA_CHECK(cudaGetLastError());
@@ -36,11 +38,13 @@ void SigmoidFocalLossBackwardCUDAKernelLauncher(Tensor input, Tensor target,
   cudaStream_t stream = at::cuda::getCurrentCUDAStream();
   AT_DISPATCH_FLOATING_TYPES_AND_HALF(
       input.scalar_type(), "sigmoid_focal_loss_backward_cuda_kernel", [&] {
-        sigmoid_focal_loss_backward_cuda_kernel<scalar_t>
-            <<<GET_BLOCKS(output_size), THREADS_PER_BLOCK, 0, stream>>>(
-                output_size, input.data_ptr<scalar_t>(),
-                target.data_ptr<int64_t>(), weight.data_ptr<scalar_t>(),
-                grad_input.data_ptr<scalar_t>(), gamma, alpha, num_classes);
+        sigmoid_focal_loss_backward_cuda_kernel<scalar_t> << <GET_BLOCKS(
+                                                                  output_size),
+                                                              THREADS_PER_BLOCK,
+                                                              0, stream>>>
+            (output_size, input.data_ptr<scalar_t>(),
+             target.data_ptr<int64_t>(), weight.data_ptr<scalar_t>(),
+             grad_input.data_ptr<scalar_t>(), gamma, alpha, num_classes);
       });
 
   AT_CUDA_CHECK(cudaGetLastError());
@@ -53,17 +57,19 @@ void SoftmaxFocalLossForwardCUDAKernelLauncher(Tensor softmax, Tensor target,
   int output_size = output.numel();
   int num_classes = softmax.size(1);
 
-  AT_ASSERTM(target.max().item<long>() <= (long)num_classes,
+  AT_ASSERTM(target.max().item<int64_t>() <= (int64_t)num_classes,
              "target label should smaller or equal than num classes");
   at::cuda::CUDAGuard device_guard(softmax.device());
   cudaStream_t stream = at::cuda::getCurrentCUDAStream();
   AT_DISPATCH_FLOATING_TYPES_AND_HALF(
       softmax.scalar_type(), "softmax_focal_loss_forward_cuda_kernel", [&] {
-        softmax_focal_loss_forward_cuda_kernel<scalar_t>
-            <<<GET_BLOCKS(output_size), THREADS_PER_BLOCK, 0, stream>>>(
-                output_size, softmax.data_ptr<scalar_t>(),
-                target.data_ptr<int64_t>(), weight.data_ptr<scalar_t>(),
-                output.data_ptr<scalar_t>(), gamma, alpha, num_classes);
+        softmax_focal_loss_forward_cuda_kernel<scalar_t> << <GET_BLOCKS(
+                                                                 output_size),
+                                                             THREADS_PER_BLOCK,
+                                                             0, stream>>>
+            (output_size, softmax.data_ptr<scalar_t>(),
+             target.data_ptr<int64_t>(), weight.data_ptr<scalar_t>(),
+             output.data_ptr<scalar_t>(), gamma, alpha, num_classes);
       });
 
   AT_CUDA_CHECK(cudaGetLastError());
@@ -79,28 +85,34 @@ void SoftmaxFocalLossBackwardCUDAKernelLauncher(Tensor softmax, Tensor target,
   int output_size = buff.numel();
   at::cuda::CUDAGuard device_guard(grad_input.device());
   cudaStream_t stream = at::cuda::getCurrentCUDAStream();
-  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
-      grad_input.scalar_type(), "softmax_focal_loss_backward_cuda1_kernel",
-      [&] {
-        softmax_focal_loss_backward_cuda1_kernel<scalar_t>
-            <<<GET_BLOCKS(output_size), THREADS_PER_BLOCK, 0, stream>>>(
-                output_size, softmax.data_ptr<scalar_t>(),
-                target.data_ptr<int64_t>(), weight.data_ptr<scalar_t>(),
-                buff.data_ptr<scalar_t>(), gamma, alpha, num_classes);
-      });
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(grad_input.scalar_type(),
+                                      "softmax_focal_loss_backward_cuda1_"
+                                      "kernel",
+                                      [&] {
+    softmax_focal_loss_backward_cuda1_kernel<scalar_t> << <GET_BLOCKS(
+                                                               output_size),
+                                                           THREADS_PER_BLOCK, 0,
+                                                           stream>>>
+        (output_size, softmax.data_ptr<scalar_t>(), target.data_ptr<int64_t>(),
+         weight.data_ptr<scalar_t>(), buff.data_ptr<scalar_t>(), gamma, alpha,
+         num_classes);
+  });
 
   AT_CUDA_CHECK(cudaGetLastError());
 
   output_size = grad_input.numel();
-  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
-      grad_input.scalar_type(), "softmax_focal_loss_backward_cuda2_kernel",
-      [&] {
-        softmax_focal_loss_backward_cuda2_kernel<scalar_t>
-            <<<GET_BLOCKS(output_size), THREADS_PER_BLOCK, 0, stream>>>(
-                output_size, softmax.data_ptr<scalar_t>(),
-                target.data_ptr<int64_t>(), buff.data_ptr<scalar_t>(),
-                grad_input.data_ptr<scalar_t>(), num_classes);
-      });
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(grad_input.scalar_type(),
+                                      "softmax_focal_loss_backward_cuda2_"
+                                      "kernel",
+                                      [&] {
+    softmax_focal_loss_backward_cuda2_kernel<scalar_t> << <GET_BLOCKS(
+                                                               output_size),
+                                                           THREADS_PER_BLOCK, 0,
+                                                           stream>>>
+        (output_size, softmax.data_ptr<scalar_t>(), target.data_ptr<int64_t>(),
+         buff.data_ptr<scalar_t>(), grad_input.data_ptr<scalar_t>(),
+         num_classes);
+  });
 
   AT_CUDA_CHECK(cudaGetLastError());
 }

